src/Symfony/Component/DependencyInjection/Compiler/DecoratorServicePass.php

```diff
@@ -46,6 +50,26 @@ public function process(ContainerBuilder $container)
             }
             $definition->innerServiceId = $renamedId;
 
+            if (!$container->has($inner)) {
+                $invalidBehavior = $decorated[3] ?? ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE;
+
+                if (ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE === $invalidBehavior) {
+                    throw new ServiceNotFoundException($inner, $id);
+                }
+
+                if (ContainerInterface::IGNORE_ON_INVALID_REFERENCE === $invalidBehavior) {
+                    $container->removeDefinition($id);
+                }
+
+                if (ContainerInterface::NULL_ON_INVALID_REFERENCE === $invalidBehavior) {
+                    $innerReferenceId = array_search($renamedId, array_map('strval', $definition->getArguments()));
```

Comment:
this doesn't work
actually, this might invalidate the entire implementation :)
what about relying on the type of reference instead of using a new attribute?